src/anomalib/metrics/per_image/utils.py

```diff
@@ -0,0 +1,413 @@
+            raise TypeError(msg)
+
+        if model_name == "":
+            msg = "Expected model name to be non-empty, but got empty string."
+            raise ValueError(msg)
+
+    num_redundant_models = len(models_ordered) - len(set(models_ordered))
+    if num_redundant_models > 0:
+        msg = f"Expected models ordered to have unique models, but got {num_redundant_models} redundant models."
+        raise ValueError(msg)
+
+
+def _validate_confidences(confidences: dict[tuple[str, str], float]) -> None:
+    if not isinstance(confidences, dict):
+        msg = f"Expected confidences to be a dict, but got {type(confidences)}."
+        raise TypeError(msg)
+
+    for (model1, model2), confidence in confidences.items():
+        if not isinstance(model1, str):
+            msg = f"Expected model name to be a string, but got {type(model1)} for model {model1}."
+            raise TypeError(msg)
+
+        if not isinstance(model2, str):
+            msg = f"Expected model name to be a string, but got {type(model2)} for model {model2}."
+            raise TypeError(msg)
+
+        if not isinstance(confidence, float):
+            msg = f"Expected confidence to be a float, but got {type(confidence)} for models {model1} and {model2}."
+            raise TypeError(msg)
+
+        if not (0 <= confidence <= 1):
+            msg = f"Expected confidence to be between 0 and 1, but got {confidence} for models {model1} and {model2}."
+            raise ValueError(msg)
+
+
+def _joint_validate_models_ordered_and_confidences(
+    models_ordered: tuple[str, ...],
+    confidences: dict[tuple[str, str], float],
+) -> None:
+    num_models = len(models_ordered)
+    expected_num_pairs = num_models * (num_models - 1)
+
+    if len(confidences) != expected_num_pairs:
+        msg = f"Expected {expected_num_pairs} pairs of models, but got {len(confidences)} pairs of models."
+        raise ValueError(msg)
+
+    models_in_confidences = {model for pair_models in confidences for model in pair_models}
+
+    diff = set(models_ordered).symmetric_difference(models_in_confidences)
+    if len(diff) > 0:
+        msg = (
+            "Expected models in confidences to be the same as models ordered, but got models missing in one"
+            f"of them: {diff}."
+        )
+        raise ValueError(msg)
+
+
+# =========================================== FUNCTIONS ===========================================
+
+
+def per_image_scores_stats(
+    per_image_scores: Tensor,
+    images_classes: Tensor | None = None,
+    only_class: int | None = None,
+    outliers_policy: str | None = StatsOutliersPolicy.NONE,
+    repeated_policy: str | None = StatsRepeatedPolicy.AVOID,
+    repeated_replacement_atol: float = 1e-2,
+) -> list[dict[str, str | int | float]]:
+    """Torch-oriented interface for `per_image_scores_stats`. See its dscription for more details (below).
+
+    Numpy version docstring
+    =======================
+
+    {docstring}
```

Comment:
it is, the documentation is built programatically from the the other function because this is just a wrapper
i wasnt sure how this would interact with the documentation builder tho
i wanted to avoid repeating documentation to avoid mistakes

Reply:
i reverted this